### PR TITLE
Add optional single-threaded feature to bevy_ecs/bevy_tasks

### LINKED ISSUE #6689: Options for disabling all parallelism for single-threaded performance
## What problem does this solve or what need does it fill?
In certain environments, like low cost VPS game hosting, it is often more efficient to host multiple single threaded game instances instead of hosting. For these use cases, the additional synchronization overhead of many `Send`/`Sync` types can be quite high, particularly with atomics.

## What solution would you like?
A feature flag on `bevy_ecs` and `bevy_tasks` to:

 - disable the `ParallelExecutor` as a default runner
 - the multi-threaded `TaskPool`
 - internally replace `Query::par_for_each` calls with `for_each`.
 - switch to using `!Send` or `!Sync` options of common types (i.e. `Mutex` -> `RefCell`, `Arc` -> `Rc`) (questionable if this is possible, we already avoid the use of these).

In these target environments, rendering is typically not required. Ideally this shouldn't require too much code change on the user's end other than changing some crate features so that code can be easily shared between client and server.

## What alternative(s) have you considered?
Leave it as is, eat the cost of atomics in these systems.

## Additional context
Original discussion: https://www.reddit.com/r/rust/comments/ytiv2a/comment/iw4q6ed/?utm_source=share&utm_medium=web2x&context=3

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Fixes #6689.

## Solution
Add `single-threaded` as an optional non-default feature to `bevy_ecs` and `bevy_tasks` that:
 
 - disable the `ParallelExecutor` as a default runner
 - disables the multi-threaded `TaskPool`
 - internally replace `QueryParIter::for_each` calls with `Query::for_each`.

Removed the `Mutex` and `Arc` usage in the single-threaded task pool.

image

## Future Work/TODO
Create type aliases for `Mutex`, `Arc` that change to single-threaaded equivalents where possible.

---

## Changelog
Added: Optional default feature `multi-theaded` to that enables multithreaded parallelism in the engine. Disabling it disables all multithreading in exchange for higher single threaded performance. Does nothing on WASM targets.

### CODE REVIEW

**Review by alice-i-cecile — COMMENTED** [review]
Okay, I think I'm sold on the feature based approach. It makes this relatively maintainable, and very easy to toggle on and off at a high-level.

**Comment by james7132** [comment]
There is some open space for discussion on whether we want it to be a feature that is turned on or off to go single threaded. `tokio` uses `rt-multi-thread` to enable multi threading and I don't think it's on by default, but we could do that instead. Requiring disabling `default-features` to switch to single threaded. It'd also make the CI job that runs with `--all-features` default to what we have now.

**Comment by alice-i-cecile** [comment]
Conceptually it feels like "multithreading" is a feature we should enable, which should be turned on by default.

We should make sure that both single-threaded and multithreaded paths are being checked in CI (and maybe in miri even): there's a lot of room for subtle mistakes to be introduced.

**Comment by recatek** [comment]
Could this also allow us to make bevy_tasks an optional dependency for bevy_ecs? The task pools are only used in a few places.

**Comment by james7132** [comment]
We could. However, this would mean that types like ParallelExeuctor straight up don't exist when single threaded is disabled, which can cause compile failures when the feature is enabled. Other parts of the engine will still rely on it since we still need an async executor for async tasks (i.e. asset loading, maybe networking).

**Comment by recatek** [comment]
Other parts of the engine rely on bevy_tasks or on the parallel executor? If it's the former, could they just pull the parts of bevy_tasks they need out of their own dependency on it?

**Comment by Guvante** [comment]
This looks in flight but I would note the current version has some sections having a `single-threaded` feature and others having a default `multi-threaded` feature, certainly unifying them would be important before merging.

**Review by cart — APPROVED** [review]
This seems like the right approach to me!

**Comment by james7132** [comment]
Will re-enable once the checks pass.

**Comment by alice-i-cecile** [comment]
I want to review this but the diff looks very messed up. Can you fix / remake this?

File: Cargo.toml [hunk 1]
```diff
@@ -199,6 +200,10 @@ filesystem_watcher = ["bevy_internal/filesystem_watcher"]
 # Enable serialization support through serde
 serialize = ["bevy_internal/serialize"]
 
+# Enables multithreaded parallelism in the engine. Disabling it forces all engine 
+# tasks to run on a single thread.
+multi-threaded = ["bevy_internal/multi-threaded"]
```
Comment (mockersf) [inline Cargo.toml:205]: I guess it's still failing because multi line is not handled correctly, could you try putting it all in one line?

Comment (james7132) [reply Cargo.toml]: Ah, good catch, wasn't understanding why it was adding the `#` there.

**Pull request merged**